### PR TITLE
MultiGet async IO across multiple levels

### PR DESCRIPTION
**Pull request by anand1976**
This PR exploits parallelism in MultiGet across levels. It applies only to the coroutine version of MultiGet. Previously, MultiGet file reads from SST files in the same level were parallelized. With this PR, MultiGet batches with keys distributed across multiple levels are read in parallel. This is accomplished by splitting the keys not present in a level (determined by bloom filtering) into a separate batch, and processing the new batch in parallel with the original batch.

Test plan:
1. Ensure existing MultiGet unit tests pass, updating them as necessary
2. New unit tests - TODO
3. Run stress test - TODO

No noticeable regression (<1%) without async IO -
Without PR: `multireadrandom :       7.261 micros/op 1101724 ops/sec 60.007 seconds 66110936 operations;  571.6 MB/s (8168992 of 8168992 found)`
With PR: `multireadrandom :       7.305 micros/op 1095167 ops/sec 60.007 seconds 65717936 operations;  568.2 MB/s (8271992 of 8271992 found)`

For a fully cached DB, but with async IO option on, no regression observed (<1%) -
Without PR: `multireadrandom :       5.201 micros/op 1538027 ops/sec 60.005 seconds 92288936 operations;  797.9 MB/s (11540992 of 11540992 found) `
With PR: `multireadrandom :       5.249 micros/op 1524097 ops/sec 60.005 seconds 91452936 operations;  790.7 MB/s (11649992 of 11649992 found) `

### CODE REVIEW

File: db/version_set.cc [hunk 1]
```diff
@@ -2380,6 +2429,181 @@ void Version::MultiGet(const ReadOptions& read_options, MultiGetRange* range,
   }
 }
 
+#ifdef USE_COROUTINES
+Status Version::ProcessBatch(
+    const ReadOptions& read_options, FilePickerMultiGet* batch,
+    std::vector<folly::coro::Task<Status>>& mget_tasks,
+    std::unordered_map<uint64_t, BlobReadContexts>* blob_ctxs,
+    autovector<FilePickerMultiGet, 4>& batches, std::queue<size_t>& waiting,
```
Comment (akankshamahajan15) [inline db/version_set.cc:2437]: What's the `waiting` queue used for?

File: db/version_set.cc [hunk 2]
```diff
@@ -2380,6 +2429,181 @@ void Version::MultiGet(const ReadOptions& read_options, MultiGetRange* range,
...
+    if (to_process.empty()) {
+      if (s.ok() && mget_tasks.size() > 0) {
+        RecordTick(db_statistics_, MULTIGET_COROUTINE_COUNT, mget_tasks.size());
+        // Collect all results so far
+        std::vector<Status> statuses = folly::coro::blockingWait(
+            folly::coro::collectAllRange(std::move(mget_tasks))
+                .scheduleOn(&range->context()->executor()));
+        for (Status stat : statuses) {
+          if (!stat.ok()) {
+            s = stat;
```
Comment (akankshamahajan15) [inline db/version_set.cc:2591]: the `s` is overriding in case of multiple failures. Is this intentional?

File: db/version_set.cc (hunk 1 above)
Comment (anand1976) [reply db/version_set.cc:2437]: Its used for an optimization on line 2489 - if there are no waiting batches, that means this is the only batch and we can possibly skip the coroutine and call the synchronous function.

File: db/version_set.cc (hunk 2 above)
Comment (anand1976) [reply db/version_set.cc:2591]: Not intentional, just copy-paste from `Version::MultiGet`. I don't have a strong opinion on what the behavior should be - we could either return the first error or the last.

**Review by akankshamahajan15 — APPROVED** [review]
LGTM!

**Pull request closed without merging**